spyder/plugins/ipythonconsole/widgets/shell.py

```diff
@@ -452,6 +453,26 @@ def change_mpl_backend(self, command):
             if not 'inline' in command:
                 self.silent_execute(command)
 
+    # ---- Spyder-kernels methods -------------------------------------------
+    def get_editor(self, filename=None):
+        """Get editor for filename and set it as the current editor."""
+        editorstack = self.editorstack()
+
+        if not filename:
+            return editorstack.get_current_editor()
+
+        index = editorstack.has_filename(filename)
+        if index is None:
+            return None
+
+        editor = editorstack.data[index].editor
+        editorstack.set_stack_index(index)
+        return editor
+
+    def handle_kernel_api_savefiles(self):
+        """Save the open files."""
+        self.editorstack().save()
```

Comment:
This is not the way we decided to handle interactions between plugins: if a widget from one plugin needs to call another plugin, it needs to emit a Qt signal, then that signal must be handled by its parent plugin, which asks the other plugin to perform the desired action.

So in this case you need to emit a signal at this point (say `sig_save_file`) with the filename you want to save. Then you need to connect that signal to a slot in `register_client` of `plugins/ipythonconsole/plugin.py` (say `save_runfile`). Since that slot is at the plugin level, it can call directly `self.main.editor` to detect if `filename` is open there, then call the current editorstack to save it.

Reply:
Are the Qt signals syncronous? The goal here is to save the file before resuming execution.

Reply:
Right, then you need to use a QEventLoop to make it synchronous.

Reply:
Does using a `QEventLoop` to make a blocking call not kind of defeat the point of using Qt signals? I mean I can either:
 - Call a method of the plugin and use the return value.
 - Emit a signal to the plugin that will then emit a signal with the reply, while I have a `QEventLoop` running, inplement a mechanism to save the reply (Another method and a class attribute?), implement a mechanism to stop the `QEventLoop` when the reply is recieved (Probably just connecting the reply signal to the Slot of `QEventLoop`), and implement a mechanism to get the reply after the `QEventLoop` finishes.

The second one seems to add quite a lot of complexity to the code. What is the adventage of this approach? I mean in general I would agree that for async calls Qt signals are superior, but I am not so sure about sync calls. Is there anywhere else in the code where a sync call is made with a `QEventLoop`? The only place I can think of is the communication with the kernel, but there a direct call was not possible.

Reply:
> The second one seems to add quite a lot of complexity to the code

Yeah, you're right. As soon as I wrote you about using a QEventLoop, I realized it makes things way too complex.

Then let's change this to the following:

```python
plugin = self.ipyclient.plugin
if plugin.main.editor is not None:
    editor = plugin.main.editor
    editorstack = editor.get_current_editorstack()
    editorstack.save()
```

For this to work you also need to create an attribute called `plugin` in `ClientWidget` that holds the arg `plugin` passed to it.